subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanPluginAutoApply.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.internal.scan.config;
+
+import com.google.common.collect.Lists;
+import org.gradle.StartParameter;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.initialization.dsl.ScriptHandler;
+import org.gradle.plugin.management.internal.DefaultPluginRequest;
+import org.gradle.plugin.management.internal.DefaultPluginRequests;
+import org.gradle.plugin.management.internal.PluginRequestInternal;
+import org.gradle.plugin.management.internal.PluginRequests;
+import org.gradle.plugin.management.internal.PluginRequestsTransformer;
+import org.gradle.plugin.use.PluginId;
+import org.gradle.plugin.use.internal.DefaultPluginId;
+import org.gradle.util.CollectionUtils;
+
+import java.util.List;
+
+/**
+ * Automatically adds a plugin request for the build scan plugin when `--scan` is detected.
+ *
+ * - Plugin request is only added for the root project.
+ * - No request is added if the project has already requested the build scan plugin, via `plugins` or `buildscript`.
+ * - The plugin request is inserted before any other plugin requests.
+ * - A fixed version of the build scan plugin is requested.
+ */
+public class BuildScanPluginAutoApply implements PluginRequestsTransformer {
+    public static final String BUILD_SCAN_PLUGIN_AUTO_APPLY_VERSION = "1.10";
```

Comment:
:o: How are we going to maintain this? Should it be part of the release process to upgrade it? Or maybe we should automate that by adding a resource, which is generated from the actual latest published release?

Reply:
👍 to what Cédric said

In any case, Kotlin DSL will need to get a hand on the version for https://github.com/gradle/kotlin-dsl/issues/490

Reply:
In practice we've seen that a new Gradle version often times requires a new build scan version. In a way both product ship together. What I'd like to see in the future is a way to expose metadata for a plugin that describes it's compatibility with a range of Gradle versions. In the short term, we could upgrade the version as part of the development process. Ultimately, the Gradleception build should fail if the build scan plugin doesn't work anymore for whatever reason.

@eskatos I guess we could provide an internal API that exposes a registry of used plugin versions. Can you explain why the Kotlin DSL needs to know about the build scan version?

Reply:
@bmuschko the Kotlin DSL provides a [type-safe accessor to apply the build scan plugin](https://github.com/gradle/kotlin-dsl/blob/410c7fa9fdd12316cb60c81e32235dfaba748c59/provider/src/main/kotlin/org/gradle/kotlin/dsl/PluginDependenciesSpecExtensions.kt#L22-L32) in the following way:
```kotlin
plugins {
    `build-scan`
}
```

Reply:
@eskatos Thanks for the info!

Reply:
I think we can address the concern separately. I created the issue https://github.com/gradle/gradle/issues/3017 for it.